Don't remove all mirror repository's releases when mirroring

lunny opened the pull request:
Fix #22066

# Purpose

This PR fix the releases will be deleted when mirror repository sync the tags. 

# The problem

In the previous implementation of #19125. All releases record in databases of one mirror repository will be deleted before sync.
Ref: https://github.com/go-gitea/gitea/pull/19125/files#diff-2aa04998a791c30e5a02b49a97c07fcd93d50e8b31640ce2ddb1afeebf605d02R481

# The Pros

This PR introduced a new method which will load all releases from databases and all tags on git data into memory. And detect which tags needs to be inserted, which tags need to be updated or deleted. Only tags releases(IsTag=true) which are not included in git data will be deleted, only tags which sha1 changed will be updated. So it will not delete any real releases include drafts.

# The Cons

The drawback is the memory usage will be higher than before if there are many tags on this repository. This PR defined a special release struct to reduce columns loaded from database to memory.

wxiaoguang reviewed (commented):
Although I was requested for reviewing, but: no test, no comment, no description, so no idea how to review.

lunny commented:
> Although I was requested for reviewing, but: no test, no comment, no description, so no idea how to review.

Sorry. I have reverted unnecessary refactor changes and also added a unit test. Please review it again.

6543 commented on modules/repository/repo_test.go:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	inserts, deletes, updates := calcSync(gitTags, dbReleases)
+	assert.EqualValues(t, 1, len(inserts), "inserts")
+	assert.EqualValues(t, *gitTags[2], *inserts[0], "inserts equal")
```
```suggestion
	if assert.EqualLen(t, inserts, 1) {
		assert.EqualValues(t, *gitTags[2], *inserts[0], "inserts equal")
	}
```

wolfogre replied:
I don't think we have to do this since it's just testing. If it panics, just let it panic.

6543 commented on modules/repository/repo_test.go:
```diff
@@ -0,0 +1,73 @@
+	assert.EqualValues(t, 1, len(inserts), "inserts")
+	assert.EqualValues(t, *gitTags[2], *inserts[0], "inserts equal")
+
+	assert.EqualValues(t, 1, len(deletes), "deletes")
```
same ...
